### ISSUE TITLE
Calibrate the compass without turning the boat over

### ISSUE THREAD

**Issue by takluyver** [Maritime-Robotics-Student-Society/sailing-robot#118]
As funny as it was to watch, trying to roll the boat through 360 degrees while keeping the IMU in place is impractical, and we would never be able to do it with the mast in place.

I think there should be two components of the 3D magnetic field we can detect: the earth's field and the combined field of components on the boat. With a smarter calibration routine, it should be possible to distinguish those without such crazy manouevering.

**Comment by takluyver** [comment]
I think I've thought of how to do this. We calculate the corrected y field (from side to side) like this:

```python
MagY_comp = (MagX*math.sin(roll)*math.sin(pitch)) +\
     (MagY*math.cos(roll)) - (MagZ*math.sin(roll)*math.cos(pitch)) 
```

If we calibrate MagX and MagY holding the boat level, as we already do, and then ask the user to point it in a constant direction and roll it from side to side, we should be able to find the coefficient and offset for MagZ that keeps `MagY_comp` closest to flat as `roll` changes.

**Comment by takluyver** [comment]
I have worked out how to do this calibration [in a notebook](https://github.com/Maritime-Robotics-Student-Society/sailing-robot/blob/master/notes/roll_calibration/Roll%20calibration.ipynb). I still need to put it into the calibration code. It will require scipy to be installed on the raspi.

**Comment by takluyver** [comment]
I have committed a new calibration script. This works on the theory I described above, and it should involve two calibration steps for the compass:

1. Holding the boat level, we turn it around horizontally. This is the calibration dance we already do, except we should now consciously try to hold it flat.
2. Keeping it pointing to a constant heading, roll from side to side as far as practical. I think tilting to 45-60 degrees each way should be enough to calibrate the Z dimension.

I have tested the calculations for this in [a notebook](https://github.com/Maritime-Robotics-Student-Society/sailing-robot/blob/master/notes/roll_calibration/Roll%20calibration.ipynb), but we still need to test the script itself.
